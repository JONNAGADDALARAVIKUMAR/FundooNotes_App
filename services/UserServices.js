import Firebase from '../config/Firebase';
import firebase from 'firebase';
import { AccessToken, LoginManager } from 'react-native-fbsdk';
import KeyChain from 'react-native-keychain';
import UserNoteServices from './UserNoteServices';
import SQLiteStorageServices from './SQLiteStorageServices';

class UserServices {

    createAccount = (email, password) => {
        return new Promise((resolve, reject) => {
            Firebase.auth().createUserWithEmailAndPassword(email, password)
            .then((user) => {
                resolve(user)
            })
            .catch((error) => {
                if (error.code === 'auth/email-already-in-use') {
                        reject('email in use!')
                }
                if (error.code === 'auth/invalid-email') {
                        reject('invalid email!')
                }
                //console.log(error.code);
            })
        })
    }

    logIn = (email, password) => {
        return new Promise((resolve, reject) => {
            Firebase.auth().signInWithEmailAndPassword(email, password)
            .then(user => {
                this.fillSQLiteDatabaseWithMissingNotesByFetchingFromFirebase()
                resolve(user)
            })
            .catch((error) => {
                if(error.code === 'auth/user-not-found') {
                    reject('User not Found')
                }
                else if(error.code === 'auth/wrong-password') {
                    reject('Invalid Password')
                }
                else if(error.code === 'auth/invalid-email') {
                    reject('Invalid Email')
                }
                //console.log(error.code);
            })
        })
    }

    resetPassword = (email) => {
        
        return new Promise((resolve, reject) => {
            Firebase.auth().sendPasswordResetEmail(email)
            .then(() => {
                resolve(true)
            })
            .catch((error) => {
                if(error.code === 'auth/invalid-email') {
                    reject('invalid email')
                } else if(error.code === 'auth/user-not-found') {
                    reject('User not found')
                }
                //console.log(error)
            })
        })
    }

    logInWithFacebook = () => {
        return new Promise(async (resolve, reject) => {
            const result = await LoginManager.logInWithPermissions(['public_profile', 'email']);
            if (result.isCancelled) {
                reject('User cancelled the login process');
            }
            const data = await AccessToken.getCurrentAccessToken();
            if (!data) {
                reject('Something went wrong obtaining access token');
            }
            const facebookCredential = await firebase.auth.FacebookAuthProvider.credential(data.accessToken);
            await firebase.auth().signInWithCredential(facebookCredential).then((user) => {
                resolve(user)
            })
        })
    }

    writeUserDataToRealTimedataBase = (email, firstName, lastName, uid) => {
        try {
            Firebase.database().ref('users/' + uid).set({
                firstName : firstName,
                lastName : lastName,
                email : email,
            })
        }
        catch(error) {
            //console.log(error);
        }
    }

    getDetails = () => {
        return new Promise(async (resolve) => {
            const user = await KeyChain.getGenericPassword();
            const userDetails = JSON.parse(user.password);
            Firebase.database().ref('users/' +userDetails.user.uid).once('value').then(async snapShot => { 
                resolve(snapShot.val())
            })
            .catch(error => console.log(error))
        })
    }

    uploadProfileImageUrlToUsers = (imageUrl) => {
        this.getDetails().then(async userDetails => {
            const user = await KeyChain.getGenericPassword();
            const Details = JSON.parse(user.password);
            Firebase.database().ref('users/' + Details.user.uid).set({
                firstName : userDetails.firstName,
                lastName : userDetails.lastName,
                email : userDetails.email,
                imageURL: imageUrl
            })
        })
    }

    logOutFromFirebase = () => {
        Firebase.auth().signOut().then(() => console.log('signed out'))
    }

    fillSQLiteDatabaseWithMissingNotesByFetchingFromFirebase = () => {
        UserNoteServices.getDetailsFromFirebase()
        .then(firebaseNotes => {
            SQLiteStorageServices.getDetailsFromSQLiteDatabase()
            .then(SQLiteNotes => {
                let NoteKeys = []
                for(let i = 0; i < SQLiteNotes.rows.length; i++) {
                    NoteKeys.push(SQLiteNotes.rows.item(i).NoteKey)
                }
                let keys = Object.keys(firebaseNotes);
                keys.map(key => {
                    if(!NoteKeys.includes(key)){
                        SQLiteStorageServices.addDetailsInSQLiteDataBaseFromFirebase(key, firebaseNotes[key].notes.title, firebaseNotes[key].notes.note, firebaseNotes[key].notes.isDeleted)
                    }
                })
            })
        })
    }
}

export default new UserServices();