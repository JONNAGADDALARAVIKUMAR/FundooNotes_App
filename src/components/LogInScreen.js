import React, { Component } from 'react';
import {Image, ScrollView, TextInput, View, Text, TouchableOpacity, Dimensions} from 'react-native';
import LogInScreenStyles from '../styles/LogInPageStyles';
import UserServices from '../../services/UserServices';
import {Button} from 'react-native-paper';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {strings} from '../Languages/strings';

export default class LogInScreen extends Component {
    constructor(props) {
        super(props);
        this.state = {
            email : '',
            password : '',
            passwordSecurity: true,
            emailError: '',
            passwordError: '',
            emailField: false,
            passwordField: false,
            isLoggedIn: false   
        }
    }
    
    async componentDidMount(){
        try {
            const isLoggedIn = JSON.parse(await AsyncStorage.getItem('isLoggedIn'))
            //console.log(isLoggedIn);
            if(isLoggedIn) {
              this.props.navigation.navigate("DashBoard")
            }
          }
          catch(e) {
            //console.log(e)
          }
    }

    emailHandler = async (enteredEmail) => {
        await this.setState({
            email: enteredEmail,
            emailField: false,
            emailError: ''
        })
    }

    passwordHandler = async (enteredPassword) => {
        await this.setState({
            password: enteredPassword,
            passwordField: false,
            passwordError: ''
        })
    }

    passwordSecurityHandler = async () => {
        const {onPress} = this.props
        var passwordSecurity = this.state.passwordSecurity
        if(passwordSecurity == true) {
            this.setState({
                passwordSecurity: false
            })
        }
        else {
            this.setState({
                passwordSecurity: true
            })
        }
        (this.props == undefined ) ? null : onPress();
    }

    handleLogInButton = async () => {
        const {onPress} = this.props
        if(this.state.email != '' && this.state.password != '') {
            await UserServices.logIn(this.state.email, this.state.password)
            .then( async (user) => {
                this.storeIteminAsyncStorage(user)
                this.props.navigation.navigate('DashBoard')
            })
            .catch(error => {
                if(error === 'User not Found') {
                    this.setState({
                        emailError: (strings.UserNotFound)
                    })
                } else if(error === 'Invalid Email') {
                    this.setState({
                        emailError: (strings.InvalidEmail)
                    })
                }   
                else if(error === 'Invalid Password') {
                    this.setState({
                        passwordError: (strings.InvalidPassword)
                    })
                }
            })
        } 
        else {
            if(this.state.email == '') {
                this.setState({
                    emailField: true
                })
            }
            if(this.state.password == '') {
                this.setState({
                    passwordField: true
                })
            }
        }
        onPress();
    }

    storeIteminAsyncStorage = async (User) => {
        try {
            await this.setState({
                isLoggedIn : true
            })
            await AsyncStorage.setItem('isLoggedIn', JSON.stringify(this.state.isLoggedIn));
        } catch (e) {
            //console.log(e);
        }
    }

    navigateToSignUpScreen = () => {
        const {onPress} = this.props
        this.props.navigation.navigate('SignUp')
        onPress();
    }

    navigateToForgotPasswordScreen = () => {
        const {onPress} = this.props
        this.props.navigation.navigate('ForgotPassword')
        onPress();
    }

    loginWithFacebook = async () => {
        UserServices.logInWithFacebook()
        .then((user) => {
            UserServices.writeUserDataToRealTimedataBase(user.additionalUserInfo.profile.email, 
                                                        user.additionalUserInfo.profile.first_name, 
                                                        user.additionalUserInfo.profile.last_name, 
                                                        user.user.uid)            
            //this.storeIteminAsyncStorage(user)
            this.props.navigation.navigate('DashBoard')
        })
        .catch((error) => {
            //console.log(error);
        })
    }

    render() {
        return (
            <View style = {LogInScreenStyles.background_Styles}>
                <ScrollView ScrollView style = {{marginBottom: '20%', marginTop: '1%'}}>
                    <Image style = {LogInScreenStyles.Logo_Style}
                        source = {require('../assets/FundoIcon.png')}
                    />
                    <Text style = {LogInScreenStyles.Fundo_Style}>{strings.AppName}</Text>
                    <TextInput
                        style = {LogInScreenStyles.Input_TextBox_Style}
                        placeholder = {strings.Email}
                        placeholderTextColor = "#b0686d"
                        onChangeText = {this.emailHandler}
                    />
                    <View style = {{flexDirection: 'row'}}>
                        <Text style = {[LogInScreenStyles.pop_up_Message, LogInScreenStyles.Fields_Missing]}>
                            {(this.state.emailField) ? (strings.EmailRequired) : null}
                        </Text>
                        <Text style = {[LogInScreenStyles.pop_up_Message, LogInScreenStyles.pop_up_Message_Flex]}>
                            {(this.state.emailError == '' || this.state.email == '') ? null : this.state.emailError}
                        </Text>
                    </View>
 
                    <View style = {[LogInScreenStyles.Input_TextBox_Style, LogInScreenStyles.set_icon]}>
                        <TextInput style = {{width: '88%'}}
                            placeholder = {strings.Password}
                            maxLength = {25}
                            placeholderTextColor = "#b0686d"
                            secureTextEntry = {this.state.passwordSecurity}
                            onChangeText = {this.passwordHandler}
                        />
                        {(this.state.passwordSecurity) 
                            ? <TouchableOpacity 
                                style= {{alignSelf: 'center'}}
                                onPress = {this.passwordSecurityHandler}>
                                    <Image style = {LogInScreenStyles.icon} source = {require('../assets/showPasswordIcon.png')}/>
                            </TouchableOpacity> 
                            : <TouchableOpacity 
                                style = {{alignSelf: 'center'}}
                                onPress = {this.passwordSecurityHandler}>
                                    <Image style = {LogInScreenStyles.icon} source = {require('../assets/hidePasswordIcon.png')}/>
                            </TouchableOpacity>  
                        }
                    </View>  

                    <View style = {{flexDirection: 'row'}}>
                        <Text style = {[LogInScreenStyles.pop_up_Message, LogInScreenStyles.Fields_Missing]}>
                            {(this.state.passwordField) ? (strings.PasswordRequired) : null}
                        </Text>
                        <Text style = {[LogInScreenStyles.pop_up_Message, LogInScreenStyles.pop_up_Message_Flex]}>
                            {(this.state.passwordError == '' || this.state.password == '') ? null : this.state.passwordError}
                        </Text>
                    </View>
                    
                    <TouchableOpacity onPress = {this.navigateToForgotPasswordScreen}>
                        <Text style = {LogInScreenStyles.forgot_Password_Style}>{strings.ForgotPassword}</Text>
                    </TouchableOpacity>

                    <View style = {{flexDirection: 'row'}}>
                        <TouchableOpacity style = {[LogInScreenStyles.LogIn_Button_Styles, LogInScreenStyles.Button_Styles]}
                            onPress = {this.handleLogInButton}>
                            <Text style = {{color: '#dbced2'}}>{strings.LogIn}</Text>
                        </TouchableOpacity>
                        
                        <TouchableOpacity style = {[LogInScreenStyles.SignUp_Button_Styles, LogInScreenStyles.Button_Styles]}
                            onPress = {this.navigateToSignUpScreen}>
                            <Text style = {{color: '#dbced2'}}>{strings.SignUp}</Text>
                        </TouchableOpacity>
                        </View>
                        <Button style = {LogInScreenStyles.Login_with_FaceBook}
                            icon = 'facebook'
                            color = 'white'
                            onPress = {this.loginWithFacebook}>
                            <Text style = {{color: '#dbced2'}}>{strings.LogInWithFacebook}</Text>
                        </Button>
                </ScrollView>
            </View>
        )
    }
}