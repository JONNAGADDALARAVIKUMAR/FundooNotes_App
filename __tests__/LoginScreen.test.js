import React from 'react';
import {shallow} from 'enzyme';
import LogInScreen from '../src/components/LogInScreen';
import UserServices from '../services/UserServices'

describe('test Login Screen', () => {
    it('test when render should match to snapshot', async () => {
        const component = shallow(<LogInScreen/>)
        expect(component).toMatchSnapshot();
    })
    it('While passing email to the emailhandler method should update email state', async () => {
        const component = shallow(<LogInScreen/>)
        expect(component.instance().state.email).toBe('')
        expect(component.instance().state.emailField).toBe(false)
        expect(component.instance().state.emailError).toBe('')
        component.instance().emailHandler('ravi@gmail.com')
        expect(component.instance().state.email).toBe('ravi@gmail.com')
    })
    it('While passing password to the  passwordhandler method should update password state', async () => {
        const component = shallow(<LogInScreen/>)
        expect(component.instance().state.password).toBe('')
        expect(component.instance().state.passwordField).toBe(false)
        expect(component.instance().state.passwordError).toBe('')
        component.instance().passwordHandler('Ravi@322')
        expect(component.instance().state.password).toBe('Ravi@322')
    })
    it('test the navigateToSignUpScreen method should Navigate to SignUp Screen', async () => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<LogInScreen onPress = {onPressEvent} navigation = {navigation}/>)
        const instance = component.instance();
        instance.navigateToSignUpScreen();

        expect(onPressEvent).toHaveBeenCalled();
        expect(navigation.navigate).toBeCalledWith("SignUp");
    })
    it('test the navigateToForgotPasswordScreen method should Navigate to ForgotPassword Screen', async () => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<LogInScreen onPress = {onPressEvent} navigation = {navigation}/>)
        const instance = component.instance();
        instance.navigateToForgotPasswordScreen();

        expect(onPressEvent).toHaveBeenCalled();
        expect(navigation.navigate).toBeCalledWith("ForgotPassword");
    })
    it('While passing email and password empty to the handleLogInButton method should update states emailField and password Field to true', async () => {
        const component = shallow(<LogInScreen/>)
        const instance = component.instance();
        instance.setState({
            email: '',
            password: '',
        })
        instance.handleLogInButton();
        expect(component.instance().state.emailField).toBe(true)
        expect(component.instance().state.passwordField).toBe(true)
    })
    it('test onPress event of eye icon of password textinput called it will change the passwordSecurity State', async () => {
        const onPressEvent = jest.fn();
        const component = shallow(<LogInScreen onPress = {onPressEvent}/>)
        const instance = component.instance();
        expect(instance.state.passwordSecurity).toBe(true);

        instance.passwordSecurityHandler();
        expect(onPressEvent).toHaveBeenCalled();
        expect(onPressEvent).toHaveBeenCalledTimes(1);
        expect(instance.state.passwordSecurity).toBe(false);

        instance.passwordSecurityHandler();
        expect(onPressEvent).toHaveBeenCalled();
        expect(onPressEvent).toHaveBeenCalledTimes(2);
        expect(instance.state.passwordSecurity).toBe(true);
    })
    
    it('test onPress event of LogIn button when passing wrong email and correct password it will return User not Found Error', async() => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<LogInScreen onPress = {onPressEvent}/>)
        const instance = component.instance();
        instance.emailHandler('ravikumarj@gmail.com')
        instance.passwordHandler('Ravi@322')
        await instance.handleLogInButton();
        expect(onPressEvent).toHaveBeenCalled();
        return UserServices.logIn(instance.state.email, instance.state.password).catch((error) => expect(instance.state.emailError).toBe('User not Found'))
    }, 10000)
    it('test onPress event of LogIn button when passing wrong email and correct password it will return Invalid Email Error', async() => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<LogInScreen onPress = {onPressEvent}/>)
        const instance = component.instance();
        instance.emailHandler('ravikumarjgmail.com')
        instance.passwordHandler('Ravi@322')
        await instance.handleLogInButton();
        expect(onPressEvent).toHaveBeenCalled();
        return UserServices.logIn(instance.state.email, instance.state.password).catch((error) => expect(instance.state.emailError).toBe('Invalid Email'))
    }, 10000)
    it('test onPress event of LogIn button when passing email and wrong password it will return Invalid Passwoard Error', async() => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<LogInScreen onPress = {onPressEvent} navigation = {navigation} />)
        const instance = component.instance();
        instance.emailHandler('ravikumarj4444@gmail.com')
        instance.passwordHandler('Ravi322')
        await instance.handleLogInButton();
        expect(onPressEvent).toHaveBeenCalled();
        return UserServices.logIn(instance.state.email, instance.state.password).catch((error) => expect(instance.state.passwordError).toBe('Invalid Password'))
    }, 10000)
    it('test onPress event of LogIn button when email and password is valid it will navigate to Dashboard Screen', async() => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<LogInScreen onPress = {onPressEvent} navigation = {navigation} />)
        const instance = component.instance();
        instance.emailHandler('ravikumarj4444@gmail.com')
        instance.passwordHandler('Ravi@322')
        await instance.handleLogInButton();
        expect(onPressEvent).toHaveBeenCalled();
        return UserServices.logIn(instance.state.email, instance.state.password).then((user) => expect(navigation.navigate).toBeCalledWith('DashBoard'))
    }, 10000)
 })