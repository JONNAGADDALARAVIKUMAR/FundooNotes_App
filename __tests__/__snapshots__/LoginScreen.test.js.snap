// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test when render should match to snapshot 1`] = `ShallowWrapper {}`;
